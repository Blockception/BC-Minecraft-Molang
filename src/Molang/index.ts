/*	Auto generated	*/
export * as Types from "./Types/index";
export * from "./Functions";
export * from "./MolangSet";
export * from "./RegExp";
export * from "./Syntax";
export * from "./Traverese";