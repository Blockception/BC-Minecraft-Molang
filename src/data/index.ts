/*	Auto generated	*/
export * from "./animations";
export * from "./animations-controllers";
export * from "./attachables";
export * from "./blocks";
export * from "./data";
export * from "./entities";
export * from "./features-rules";
export * from "./general";
export * from "./items";
export * from "./molang-data";
export * from "./molang-function";
export * from "./particles";
export * from "./render-controllers";

