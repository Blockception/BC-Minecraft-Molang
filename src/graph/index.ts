export * as Mermaid from './mermaid'