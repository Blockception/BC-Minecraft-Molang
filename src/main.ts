export * as MolangData from "./MolangData/include";
export * from "./Types/Defined Using/include";
export * from "./Molang/include";
