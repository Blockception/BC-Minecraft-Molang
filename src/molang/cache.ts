import { OffsetWord } from "bc-minecraft-bedrock-types/lib/types";
import { ExpressionNode, parseMolang } from "./syntax";

export class MolangSyntaxCache {
  private _data: Map<string, ExpressionNode[]>;

  constructor() {
    this._data = new Map();
  }

  build(code: OffsetWord): ExpressionNode[] | undefined {
    const exp = this._data.get(code.text);
    if (exp === undefined) {
      const r = parseMolang(code);
      this._data.set(code.text, r);
      return r;
    }

    return exp;
  }

  clear() {
    this._data.clear();
  }

  entries() {
    return this._data.entries();
  }

  syntaxes() {
    return this._data.keys();
  }

  expressions() {
    return this._data.values();
  }
}
