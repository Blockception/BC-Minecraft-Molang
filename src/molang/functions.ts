/* eslint-disable no-fallthrough */
/** The type of molang */
export enum MolangType {
  /** A command */
  command,
  /** An event */
  event,
  /** regular molang */
  molang,
  /** unknown */
  unknown,
}

const eventRegex = /^@s ([\w:_-]+)/im;
const commandRegex = /^\/[a-z]+/im;
const molangRegexp =
  /\b((query|math|variable|texture|temp|geometry|material|array|context|c|q|v|t)\.[A-Za-z_0-9]+|->|this)\b/im;

/**
 * Checks if the string is a molang type
 * @param data The string to check
 * @returns The type of molang
 */
export function IsMolangType(data: string): MolangType {
  if (commandRegex.test(data)) return MolangType.command;
  if (eventRegex.test(data)) return MolangType.event;
  if (molangRegexp.test(data)) return MolangType.molang;

  return MolangType.unknown;
}

/**
 * Already assumes that the given data is of type MolangType.event
 * @param data The text to parse
 */
export function getEvent(data: string): string {
  return data.slice(3).trim();
}

/**
 * Checks if the given data is a valid molang expression
 * @param molang The molang expression to check
 * @returns True if the given molang expression is valid
 */
export function isValidMolang(molang: string): boolean {
  let instr = false;
  let level = 0;

  for (let I = 0; I < molang.length; I++) {
    switch (molang[I]) {
      case "'":
        instr = !instr;

      case "[":
      case "{":
      case "(":
        if (!instr) level++;
        break;

      case "]":
      case ")":
      case "}":
        if (!instr) level--;
        break;
    }
  }

  return level === 0 && instr === false;
}

/**
 * Finds the specific molang expression in the given text
 * @param molang The text to parse
 * @param startIndex The index to start searching from
 * @param find The molang expression to find
 * @returns The index of the molang expression
 */
export function find(molang: string, startIndex: number, find: string): number {
  let instr = false;
  let level = 0;
  const length = find.length;
  const max = molang.length - length;

  for (let I = startIndex; I < max; I++) {
    if (level === 0 && instr === false) {
      if (molang.slice(I, I + length) === find) {
        return I;
      }
    }

    switch (molang[I]) {
      case "'":
        instr = !instr;

      case "[":
      case "{":
      case "(":
        if (!instr) level++;
        break;

      case "]":
      case ")":
      case "}":
        if (!instr) level--;
        break;
    }
  }

  return -1;
}
