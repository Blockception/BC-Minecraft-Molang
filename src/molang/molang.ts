/** */
export namespace molang {
  /**
   *
   * @param value
   * @returns
   */
  export function is(value: any): value is string {
    if (value) {
      if (typeof value === "string") {
        return true;
      }
    }

    return false;
  }

  /**
   *
   * @param data
   * @returns
   */
  export function getVariablesUsed(data: string): string[] {
    let matches = data.match(/([Vv]|[Vv]ariable)\.([a-zA-z0-9_]+)/gi);

    let out: string[] = [];

    if (matches) {
      for (let I = 0; I < matches.length; I++) {
        const element = matches[I];

        const index = element.indexOf(".");
        if (index > -1) {
          out.push(element.substring(index + 1, element.length));
        }
      }
    }

    return out;
  }

  /**
   *
   * @param data
   * @returns
   */
  export function getVariableDefined(data: string): string | undefined {
    let match = data.match(/^[ \t]*([Vv]ariable|[Vv])\.([a-zA-Z0-9_]+)/gi);

    if (match) {
      if (match.length >= 2) {
        return match[1];
      }
    }

    return undefined;
  }

  /**
   *
   * @param data
   * @returns
   */
  export function getQueriesUsed(data: string): string[] {
    let matches = data.match(/([Qq]uery|[Qq])\.([a-zA-Z0-9_]+)/gi);

    let out: string[] = [];

    if (matches) {
      for (let I = 0; I < matches.length; I++) {
        const element = matches[I];

        const index = element.indexOf(".");
        if (index > -1) {
          out.push(element.substring(index + 1, element.length));
        }
      }
    }

    return out;
  }
}
