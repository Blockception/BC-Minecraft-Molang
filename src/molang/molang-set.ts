import { OffsetWord } from "bc-minecraft-bedrock-types/lib/types";
import { MolangSyntaxCache } from "./cache";
import { isMolang, isValidMolang } from "./functions";
import { ExpressionNode, FunctionCallNode, NodeType, ResourceReferenceNode, VariableNode, walk } from "./syntax";

/** The interface for the molang set */
export class MolangSet {
  public cache = new MolangSyntaxCache();
  public assigned = new Set<ResourceReferenceNode | VariableNode>();
  public functions = new Set<FunctionCallNode>();
  public using = new Set<ResourceReferenceNode | VariableNode>();

  constructor() {}

  /**
   * adds the data from the molang code if it is valid molang
   * @param molang
   */
  addIf(molang: OffsetWord) {
    if (isValidMolang(molang.text)) this.add(molang);
  }

  /**
   *
   * @param molang
   * @returns
   */
  add(molang: OffsetWord) {
    const exp = this.cache.build(molang);
    if (exp === undefined) return;
    exp.forEach((e) => walk(e, this.walkFn.bind(this)));
  }

  private walkFn(node: ExpressionNode): void {
    switch (node.type) {
      case NodeType.Assignment:
        this.checkAssigned(node.left);
        break;
      case NodeType.FunctionCall:
        this.functions.add(node);
        break;
      case NodeType.ResourceReference:
      case NodeType.Variable:
        if (this.assigned.has(node)) break;
        this.using.add(node);
        break;
    }
  }

  private checkAssigned(node: ExpressionNode): void {
    switch (node.type) {
      case NodeType.ResourceReference:
      case NodeType.Variable:
        this.assigned.add(node);
        break;
    }
  }

  harvest(object: Record<string, any> | string, originalText: string): this {
    if (typeof object === "string") {
      if (isMolang(object)) {
        this.add(OffsetWord.create(object, originalText.indexOf(object)));
        return this;
      }
    }

    for (const [, value] of Object.entries(object)) {
      if (typeof value === "string") {
        if (isMolang(value)) {
          this.add(OffsetWord.create(value, originalText.indexOf(value)));
        }
      } else if (typeof value === "object") {
        if (Array.isArray(value)) {
          value.forEach((v) => this.harvest(v, originalText));
        } else {
          this.harvest(value, originalText);
        }
      }
    }

    return this;
  }

  static harvest(object: Record<string, any> | string, originalText: string): MolangSet {
    return new MolangSet().harvest(object, originalText);
  }
}
