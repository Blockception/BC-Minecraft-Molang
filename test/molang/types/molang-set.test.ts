import { MolangFullSet, MolangSet } from "../../../src/molang";
import { VanillaPlayer } from "../../vanilla-player";

describe("Molang", () => {
  describe("MolangFullSet", () => {
    const data1 = MolangFullSet.harvest(VanillaPlayer.Data);
    const data2 = MolangFullSet.harvest(VanillaPlayer.DataOBject);
    expect([data1, data2]).toMatchSnapshot();
  });

  describe("MolangSet", () => {
    const data1 = MolangSet.harvest(VanillaPlayer.Data);
    const data2 = MolangSet.harvest(VanillaPlayer.DataOBject);
    expect([data1, data2]).toMatchSnapshot();
  });

  it("MolangFullSet2", () => {
    const data = MolangFullSet.harvest(VanillaPlayer.Data);
    expect(data).toMatchSnapshot();
  });

  it("MolangFullSet", () => {
    const data = MolangSet.harvest(VanillaPlayer.Data);
    expect(data).toMatchSnapshot();
  });
});
