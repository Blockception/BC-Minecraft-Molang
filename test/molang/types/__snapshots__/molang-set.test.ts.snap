// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
[
  {
    "contexts": {
      "using": [
        "world_x",
      ],
    },
    "geometries": {
      "defined": [],
      "using": [
        "humanoid.custom",
        "cape",
      ],
    },
    "materials": {
      "defined": [],
      "using": [],
    },
    "queries": {
      "using": [
        "modified_distance_moved",
        "modified_move_speed",
        "life_time",
        "is_on_ground",
        "is_alive",
        "position_delta",
        "main_hand_item_use_duration",
        "main_hand_item_max_duration",
      ],
    },
    "temps": {
      "defined": [
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
      ],
      "using": [
        "foo",
      ],
    },
    "textures": {
      "defined": [],
      "using": [],
    },
    "variables": {
      "defined": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "tcos0",
        "first_person_rotation_factor",
        "map_angle",
        "item_use_normalized",
      ],
      "using": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "attack_time",
        "tcos0",
        "gliding_speed_value",
        "first_person_rotation_factor",
        "map_angle",
        "player_x_rotation",
        "item_use_normalized",
        "is_first_person",
        "map_face_icon",
      ],
    },
  },
  {
    "contexts": {
      "using": [
        "world_x",
      ],
    },
    "geometries": {
      "defined": [],
      "using": [
        "humanoid.custom",
        "cape",
      ],
    },
    "materials": {
      "defined": [],
      "using": [],
    },
    "queries": {
      "using": [
        "modified_distance_moved",
        "modified_move_speed",
        "life_time",
        "is_on_ground",
        "is_alive",
        "position_delta",
        "main_hand_item_use_duration",
        "main_hand_item_max_duration",
      ],
    },
    "temps": {
      "defined": [
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
      ],
      "using": [
        "foo",
      ],
    },
    "textures": {
      "defined": [],
      "using": [],
    },
    "variables": {
      "defined": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "tcos0",
        "first_person_rotation_factor",
        "map_angle",
        "item_use_normalized",
      ],
      "using": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "attack_time",
        "tcos0",
        "gliding_speed_value",
        "first_person_rotation_factor",
        "map_angle",
        "player_x_rotation",
        "item_use_normalized",
        "is_first_person",
        "map_face_icon",
      ],
    },
  },
]
`;

exports[` 2`] = `
[
  {
    "contexts": {
      "using": [
        "world_x",
      ],
    },
    "queries": {
      "using": [
        "modified_distance_moved",
        "modified_move_speed",
        "life_time",
        "is_on_ground",
        "is_alive",
        "position_delta",
        "main_hand_item_use_duration",
        "main_hand_item_max_duration",
      ],
    },
    "temps": {
      "defined": [
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
      ],
      "using": [
        "foo",
      ],
    },
    "variables": {
      "defined": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "tcos0",
        "first_person_rotation_factor",
        "map_angle",
        "item_use_normalized",
      ],
      "using": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "attack_time",
        "tcos0",
        "gliding_speed_value",
        "first_person_rotation_factor",
        "map_angle",
        "player_x_rotation",
        "item_use_normalized",
        "is_first_person",
        "map_face_icon",
      ],
    },
  },
  {
    "contexts": {
      "using": [
        "world_x",
      ],
    },
    "queries": {
      "using": [
        "modified_distance_moved",
        "modified_move_speed",
        "life_time",
        "is_on_ground",
        "is_alive",
        "position_delta",
        "main_hand_item_use_duration",
        "main_hand_item_max_duration",
      ],
    },
    "temps": {
      "defined": [
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
      ],
      "using": [
        "foo",
      ],
    },
    "variables": {
      "defined": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "tcos0",
        "first_person_rotation_factor",
        "map_angle",
        "item_use_normalized",
      ],
      "using": [
        "is_holding_right",
        "is_blinking",
        "last_blink_time",
        "hand_bob",
        "helmet_layer_visible",
        "leg_layer_visible",
        "boot_layer_visible",
        "chest_layer_visible",
        "attack_body_rot_y",
        "attack_time",
        "tcos0",
        "gliding_speed_value",
        "first_person_rotation_factor",
        "map_angle",
        "player_x_rotation",
        "item_use_normalized",
        "is_first_person",
        "map_face_icon",
      ],
    },
  },
]
`;

exports[`Molang MolangFullSet 1`] = `
{
  "contexts": {
    "using": [
      "world_x",
    ],
  },
  "queries": {
    "using": [
      "modified_distance_moved",
      "modified_move_speed",
      "life_time",
      "is_on_ground",
      "is_alive",
      "position_delta",
      "main_hand_item_use_duration",
      "main_hand_item_max_duration",
    ],
  },
  "temps": {
    "defined": [
      "helmet_layer_visible",
      "leg_layer_visible",
      "boot_layer_visible",
      "chest_layer_visible",
    ],
    "using": [
      "foo",
    ],
  },
  "variables": {
    "defined": [
      "is_holding_right",
      "is_blinking",
      "last_blink_time",
      "hand_bob",
      "helmet_layer_visible",
      "leg_layer_visible",
      "boot_layer_visible",
      "chest_layer_visible",
      "attack_body_rot_y",
      "tcos0",
      "first_person_rotation_factor",
      "map_angle",
      "item_use_normalized",
    ],
    "using": [
      "is_holding_right",
      "is_blinking",
      "last_blink_time",
      "hand_bob",
      "helmet_layer_visible",
      "leg_layer_visible",
      "boot_layer_visible",
      "chest_layer_visible",
      "attack_body_rot_y",
      "attack_time",
      "tcos0",
      "gliding_speed_value",
      "first_person_rotation_factor",
      "map_angle",
      "player_x_rotation",
      "item_use_normalized",
      "is_first_person",
      "map_face_icon",
    ],
  },
}
`;

exports[`Molang MolangFullSet2 1`] = `
{
  "contexts": {
    "using": [
      "world_x",
    ],
  },
  "geometries": {
    "defined": [],
    "using": [
      "humanoid.custom",
      "cape",
    ],
  },
  "materials": {
    "defined": [],
    "using": [],
  },
  "queries": {
    "using": [
      "modified_distance_moved",
      "modified_move_speed",
      "life_time",
      "is_on_ground",
      "is_alive",
      "position_delta",
      "main_hand_item_use_duration",
      "main_hand_item_max_duration",
    ],
  },
  "temps": {
    "defined": [
      "helmet_layer_visible",
      "leg_layer_visible",
      "boot_layer_visible",
      "chest_layer_visible",
    ],
    "using": [
      "foo",
    ],
  },
  "textures": {
    "defined": [],
    "using": [],
  },
  "variables": {
    "defined": [
      "is_holding_right",
      "is_blinking",
      "last_blink_time",
      "hand_bob",
      "helmet_layer_visible",
      "leg_layer_visible",
      "boot_layer_visible",
      "chest_layer_visible",
      "attack_body_rot_y",
      "tcos0",
      "first_person_rotation_factor",
      "map_angle",
      "item_use_normalized",
    ],
    "using": [
      "is_holding_right",
      "is_blinking",
      "last_blink_time",
      "hand_bob",
      "helmet_layer_visible",
      "leg_layer_visible",
      "boot_layer_visible",
      "chest_layer_visible",
      "attack_body_rot_y",
      "attack_time",
      "tcos0",
      "gliding_speed_value",
      "first_person_rotation_factor",
      "map_angle",
      "player_x_rotation",
      "item_use_normalized",
      "is_first_person",
      "map_face_icon",
    ],
  },
}
`;
